File: springboot/src/main/java/com/trading/engine/TradeEngine.java
package com.trading.engine;

import com.trading.config.AppConfig;
import com.trading.model.*;
import com.trading.service.DatabaseService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Service
public class TradeEngine {

    private final AppState state = AppState.get();

    @Autowired
    private DatabaseService dbService;

    // ── Market hours ──────────────────────────────────────────────────────────────

    public boolean isMarketOpen() {
        LocalTime now   = nowIST().toLocalTime();
        LocalTime open  = LocalTime.of(AppConfig.MARKET_OPEN_HOUR,  AppConfig.MARKET_OPEN_MIN);
        LocalTime close = LocalTime.of(AppConfig.MARKET_CLOSE_HOUR, AppConfig.MARKET_CLOSE_MIN);
        return !now.isBefore(open) && !now.isAfter(close);
    }

    public boolean isInTimeWindow() {
        LocalTime now   = nowIST().toLocalTime();
        LocalTime start = LocalTime.of(AppConfig.ENTRY_START_HOUR, AppConfig.ENTRY_START_MIN);
        LocalTime end   = LocalTime.of(AppConfig.ENTRY_END_HOUR,   AppConfig.ENTRY_END_MIN);
        return !now.isBefore(start) && now.isBefore(end);
    }

    public static LocalDateTime nowIST() {
        return LocalDateTime.now(ZoneId.of("Asia/Kolkata"));
    }

    // ── Leader momentum ───────────────────────────────────────────────────────────

    public LeaderResult leadersMomentum() {
        for (String stockName : AppConfig.LEADER_STOCKS) {
            AppConfig.Stock stock = AppConfig.STOCKS.stream()
                .filter(s -> s.name().equals(stockName)).findFirst().orElse(null);
            if (stock == null) continue;
            List<Candle> candles = state.lastNCandles.get(stock.symbol());
            if (candles == null || candles.isEmpty())
                return new LeaderResult("Nobuysell", "Leader candle missing: " + stockName);
        }
        List<String> buyLeaders = new ArrayList<>(), sellLeaders = new ArrayList<>();
        for (String stockName : AppConfig.LEADER_STOCKS) {
            AppConfig.Stock stock = AppConfig.STOCKS.stream()
                .filter(s -> s.name().equals(stockName)).findFirst().orElse(null);
            if (stock == null) continue;
            List<Candle> candles = state.lastNCandles.get(stock.symbol());
            if (candles == null || candles.isEmpty()) continue;
            Candle c = candles.get(candles.size() - 1);
            if (c.close > c.open) buyLeaders.add(stockName);
            else if (c.close < c.open) sellLeaders.add(stockName);
        }
        if (buyLeaders.size() >= AppConfig.SAME_DIRECTION_REQUIRED)
            return new LeaderResult("BUY",
                String.join("+", buyLeaders.subList(0, AppConfig.SAME_DIRECTION_REQUIRED)) + " aligned");
        if (sellLeaders.size() >= AppConfig.SAME_DIRECTION_REQUIRED)
            return new LeaderResult("SELL",
                String.join("+", sellLeaders.subList(0, AppConfig.SAME_DIRECTION_REQUIRED)) + " aligned");
        return new LeaderResult("Nobuysell", "No leader match");
    }

    // ── Entry check (runs on indicator thread pool) ────────────────────────────────

    @Async("indicatorExecutor")
    public void checkTradeEntryAsync() {
        try { checkTradeEntry(); } catch (Exception e) {
            System.err.println("Entry check error: " + e.getMessage());
        }
    }

    public synchronized AppState.EntryDiagnostics checkTradeEntry() {
        String interval = state.selectedInterval;
        List<Candle> bnCandles = state.lastNCandles.get(AppConfig.INDEX_SYMBOL);
        Candle bn = (bnCandles != null && !bnCandles.isEmpty())
            ? bnCandles.get(bnCandles.size() - 1) : null;

        boolean marketOpen    = isMarketOpen();
        boolean timeOk        = isInTimeWindow();
        boolean noActiveTrade = (state.activeTrade == null);
        long    cooldownMs    = System.currentTimeMillis() - state.lastExitTime;

        Double sideRange = IndicatorEngine.sidewaysRange(bnCandles);
        boolean rangeOk = sideRange != null && sideRange >= 12;
        IndicatorEngine.MomentumResult mom = IndicatorEngine.strongMomentum(bnCandles, interval);
        LeaderResult leaderSig = leadersMomentum();
        boolean sigOk = leaderSig.signal.equals("BUY") || leaderSig.signal.equals("SELL");

        int green = 0, red = 0, strongQty = 0;
        List<AppState.StockStat> stockStats = new ArrayList<>();
        for (String stockName : AppConfig.LEADER_STOCKS) {
            AppConfig.Stock stock = AppConfig.STOCKS.stream()
                .filter(s -> s.name().equals(stockName)).findFirst().orElse(null);
            if (stock == null) continue;
            List<Candle> sc = state.lastNCandles.get(stock.symbol());
            Candle c = (sc != null && !sc.isEmpty()) ? sc.get(sc.size() - 1) : null;
            if (c == null) continue;
            if (c.close > c.open) green++; else if (c.close < c.open) red++;
            double qty       = state.latestMinuteQty.getOrDefault(stockName, 0.0);
            double threshold = effectiveThreshold(stockName, interval);
            if (qty >= threshold) strongQty++;
            stockStats.add(new AppState.StockStat(stockName, c, qty, threshold));
        }

        boolean dirOk         = Math.max(green, red) >= AppConfig.SAME_DIRECTION_REQUIRED;
        boolean sqOk          = strongQty >= AppConfig.SAME_DIRECTION_REQUIRED;
        boolean candleCloseOk = bn != null && isCandleClosed(bn.startTime, interval);
        boolean alreadyTraded = bn != null && bn.startTime.substring(0, 16).equals(state.lastTradeCandle);

        BNIndicators bnInd = IndicatorEngine.checkBNIndicators();
        state.bnIndicators = bnInd;

        boolean macdMet = bnInd.macdDir != null && !bnInd.macdDir.equals("—") && !bnInd.macdDir.equals("NEUTRAL");
        boolean emaMet  = bnInd.emaStack != null && (bnInd.emaStack.bullish || bnInd.emaStack.bearish);
        boolean gateOk  = bnInd.bullish || bnInd.bearish;

        // Compute time at which current candle closes
        String candleCloseTime = null;
        if (bn != null && bn.startTime != null && !bn.startTime.isEmpty()) {
            try {
                LocalDateTime start = LocalDateTime.parse(bn.startTime.substring(0, 19),
                    DateTimeFormatter.ISO_LOCAL_DATE_TIME);
                int mins = switch (interval) { case "3m" -> 3; case "5m" -> 5; case "15m" -> 15; default -> 1; };
                candleCloseTime = start.plusMinutes(mins).format(DateTimeFormatter.ofPattern("HH:mm"));
            } catch (Exception ignored) {}
        }

        AppState.EntryDiagnostics diag = new AppState.EntryDiagnostics(
            marketOpen, timeOk, noActiveTrade, cooldownMs, sideRange, candleCloseOk,
            leaderSig.signal, leaderSig.reason, green, red, strongQty,
            alreadyTraded, bnInd, bn, stockStats,
            new AppState.MomResult(mom.ok(), mom.reason()),
            candleCloseTime
        );
        state.entryDiagnostics = diag;

        boolean allOk = marketOpen && timeOk && noActiveTrade
            && cooldownMs >= 60_000 && rangeOk && mom.ok()
            && sigOk && dirOk && sqOk && candleCloseOk
            && !alreadyTraded && macdMet && emaMet && gateOk;

        boolean allButClose = marketOpen && timeOk && noActiveTrade
            && cooldownMs >= 60_000 && rangeOk && mom.ok()
            && sigOk && dirOk && sqOk && !alreadyTraded && macdMet && emaMet && gateOk;

        if (allButClose && !candleCloseOk)
            state.pendingSignal = new AppState.PendingSignal(leaderSig.signal, leaderSig.reason);
        else
            state.pendingSignal = null;

        if (allOk && state.activeTrade == null) {
            // Use live LTP for entry — more current than stale candle close price
            double entryPrice = state.bnLTP > 0 ? state.bnLTP : bn.close;
            if (bnInd.bullish && leaderSig.signal.equals("BUY"))
                enterTrade("BUY",  entryPrice, bn.startTime.substring(0, 16), "AUTO");
            else if (bnInd.bearish && leaderSig.signal.equals("SELL"))
                enterTrade("SELL", entryPrice, bn.startTime.substring(0, 16), "AUTO");
        }
        return diag;
    }

    private boolean isCandleClosed(String startTime, String interval) {
        if (startTime == null || startTime.isEmpty()) return false;
        try {
            LocalDateTime start = LocalDateTime.parse(startTime.substring(0, 19),
                DateTimeFormatter.ISO_LOCAL_DATE_TIME);
            int mins = switch (interval) { case "3m" -> 3; case "5m" -> 5; case "15m" -> 15; default -> 1; };
            return nowIST().isAfter(start.plusMinutes(mins));
        } catch (Exception e) { return false; }
    }

    // ── Position sizing ───────────────────────────────────────────────────────────

    // 1% of available funds as risk budget; min 1 lot, max 10 lots
    private int calcNumLots() {
        double riskBudget = state.availableFunds * 0.01;
        double riskPerLot = AppConfig.STOPLOSS * AppConfig.LOT_SIZE; // 18 * 30 = 540
        int lots = (int) Math.floor(riskBudget / riskPerLot);
        return Math.max(1, Math.min(lots, 10));
    }

    // ── Enter trade ───────────────────────────────────────────────────────────────

    public void enterTrade(String type, double price, String candleTime, String confidence) {
        if (state.activeTrade != null) return;
        int numLots = calcNumLots();
        state.activeTrade     = new ActiveTrade(type, price, candleTime, confidence, numLots);
        state.lastTradeCandle = candleTime;
        state.signalLocked    = true;

        Trade t = new Trade();
        t.type       = type;
        t.price      = price;
        t.time       = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
        t.confidence = confidence;
        t.pnl        = 0;
        dbService.saveTrade(t);
        System.out.printf("ENTRY %s @ %.2f [%s] conf=%s lots=%d%n", type, price, candleTime, confidence, numLots);
    }

    // ── Exit check ────────────────────────────────────────────────────────────────

    public void checkExit() {
        if (state.activeTrade == null) return;
        List<Candle> bnCandles = state.lastNCandles.get(AppConfig.INDEX_SYMBOL);
        if (bnCandles == null || bnCandles.isEmpty()) return;
        // Use live LTP; fall back to candle close if LTP not yet received
        double price = state.bnLTP > 0 ? state.bnLTP : bnCandles.get(bnCandles.size() - 1).close;
        ActiveTrade at = state.activeTrade;
        double pnl;
        if (at.type.equals("BUY")) {
            pnl = price - at.entry;
            if (pnl >= AppConfig.TRAIL_TRIGGER)
                at.currentSL = Math.max(at.currentSL, price - AppConfig.TRAIL_DISTANCE);
            else if (pnl >= AppConfig.BREAKEVEN_TRIGGER)
                at.currentSL = Math.max(at.currentSL, at.entry);
            if (price >= at.entry + AppConfig.TARGET || price <= at.currentSL) exitTrade(price, pnl);
        } else {
            pnl = at.entry - price;
            if (pnl >= AppConfig.TRAIL_TRIGGER)
                at.currentSL = Math.min(at.currentSL, price + AppConfig.TRAIL_DISTANCE);
            else if (pnl >= AppConfig.BREAKEVEN_TRIGGER)
                at.currentSL = Math.min(at.currentSL, at.entry);
            if (price <= at.entry - AppConfig.TARGET || price >= at.currentSL) exitTrade(price, pnl);
        }
    }

    public void exitTrade(double exitPrice, double pnlPts) {
        if (state.activeTrade == null) return;
        pnlPts = Math.round(pnlPts * 100.0) / 100.0;
        int    numLots = state.activeTrade.numLots;
        double pnlRs   = Math.round(pnlPts * numLots * AppConfig.LOT_SIZE * 100.0) / 100.0;

        Trade t = new Trade();
        t.type       = state.activeTrade.type + "_EXIT";
        t.price      = exitPrice;
        t.time       = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
        t.confidence = state.activeTrade.confidence;
        t.pnl        = pnlRs;   // stored in rupees
        dbService.saveTrade(t);
        state.availableFunds += pnlRs;
        state.activeTrade   = null;
        state.lastExitTime  = System.currentTimeMillis();
        state.signalLocked  = false;
        state.pendingSignal = null;
        System.out.printf("EXIT @ %.2f PnL=%.2f pts = ₹%.2f (%d lots)%n", exitPrice, pnlPts, pnlRs, numLots);
    }

    // ── Manual entry/exit ─────────────────────────────────────────────────────────

    public void manualEntry(String type, double price) {
        if (state.activeTrade != null) return;
        enterTrade(type, price, nowIST().toString().substring(0, 16), "MANUAL");
    }

    public void manualExit() {
        if (state.activeTrade == null) return;
        List<Candle> bnCandles = state.lastNCandles.get(AppConfig.INDEX_SYMBOL);
        if (bnCandles == null || bnCandles.isEmpty()) return;
        double price = bnCandles.get(bnCandles.size() - 1).close;
        double pnl   = state.activeTrade.type.equals("BUY") ? price - state.activeTrade.entry
                                                             : state.activeTrade.entry - price;
        exitTrade(price, pnl);
    }

    // ── Helpers ───────────────────────────────────────────────────────────────────

    public static int effectiveThreshold(String stock, String interval) {
        Integer base = AppConfig.STOCK_QTY_THRESHOLD.get(stock);
        if (base == null) return 9999;
        double mult = switch (interval) { case "3m" -> 1.5; case "5m" -> 2.0; case "15m" -> 6.0; default -> 1.0; };
        return (int)(base * mult);
    }

    public record LeaderResult(String signal, String reason) {}
}

File: app/engine/trade_engine.py
from __future__ import annotations

import asyncio
import time
from datetime import datetime, timedelta, time as dtime
from typing import TYPE_CHECKING, Optional
from zoneinfo import ZoneInfo

import app.config as cfg
from app.models import ActiveTrade, Trade
from app.state import AppState, EntryDiagnostics, MomResult, PendingSignal, StockStat, get_state

if TYPE_CHECKING:
    from app.services.database import DatabaseService

IST = ZoneInfo("Asia/Kolkata")

_entry_lock: Optional[asyncio.Lock] = None


def _get_entry_lock() -> asyncio.Lock:
    global _entry_lock
    if _entry_lock is None:
        _entry_lock = asyncio.Lock()
    return _entry_lock


def now_ist() -> datetime:
    return datetime.now(IST)


def is_market_open() -> bool:
    t     = now_ist().time()
    open_ = dtime(cfg.MARKET_OPEN_HOUR,  cfg.MARKET_OPEN_MIN)
    close = dtime(cfg.MARKET_CLOSE_HOUR, cfg.MARKET_CLOSE_MIN)
    return open_ <= t <= close


def is_in_time_window() -> bool:
    t     = now_ist().time()
    start = dtime(cfg.ENTRY_START_HOUR, cfg.ENTRY_START_MIN)
    end   = dtime(cfg.ENTRY_END_HOUR,   cfg.ENTRY_END_MIN)
    return start <= t < end


def effective_threshold(stock: str, interval: str) -> int:
    base = cfg.STOCK_QTY_THRESHOLD.get(stock)
    if base is None:
        return 9999
    mult = {"3m": 1.5, "5m": 2.0, "15m": 6.0}.get(interval, 1.0)
    return int(base * mult)


def _is_candle_closed(start_time: str, interval: str) -> bool:
    if not start_time:
        return False
    try:
        s     = start_time[:19].replace(" ", "T")
        start = datetime.fromisoformat(s).replace(tzinfo=IST)
        mins  = {"3m": 3, "5m": 5, "15m": 15}.get(interval, 1)
        return now_ist() > start + timedelta(minutes=mins)
    except Exception:
        return False


def _calc_num_lots(state: AppState) -> int:
    risk_budget = state.available_funds * 0.01
    risk_per_lot = cfg.STOPLOSS * cfg.LOT_SIZE   # 18 × 30 = 540
    lots = int(risk_budget / risk_per_lot)
    return max(1, min(lots, 10))


# ── Leader momentum ───────────────────────────────────────────────────────────

def leaders_momentum():
    from app.engine.trade_engine import _LeaderResult
    st = get_state()
    for stock_name in cfg.LEADER_STOCKS:
        stock   = next((s for s in cfg.STOCKS if s.name == stock_name), None)
        candles = st.last_n_candles.get(stock.symbol, []) if stock else []
        if not candles:
            return _LeaderResult("Nobuysell", f"Leader candle missing: {stock_name}")
    buy_leaders  = []
    sell_leaders = []
    for stock_name in cfg.LEADER_STOCKS:
        stock   = next((s for s in cfg.STOCKS if s.name == stock_name), None)
        candles = st.last_n_candles.get(stock.symbol, []) if stock else []
        if not candles:
            continue
        c = candles[-1]
        if   c.close > c.open: buy_leaders.append(stock_name)
        elif c.close < c.open: sell_leaders.append(stock_name)
    n = cfg.SAME_DIRECTION_REQUIRED
    if len(buy_leaders)  >= n:
        return _LeaderResult("BUY",  "+".join(buy_leaders[:n])  + " aligned")
    if len(sell_leaders) >= n:
        return _LeaderResult("SELL", "+".join(sell_leaders[:n]) + " aligned")
    return _LeaderResult("Nobuysell", "No leader match")


class _LeaderResult:
    def __init__(self, signal: str, reason: str):
        self.signal = signal
        self.reason = reason


# ── Entry check ───────────────────────────────────────────────────────────────

async def check_trade_entry(db: "DatabaseService | None" = None) -> EntryDiagnostics:
    from app.engine.indicator_engine import (
        check_bn_indicators, sideways_range, strong_momentum
    )
    st = get_state()
    async with _get_entry_lock():
        interval   = st.selected_interval
        bn_candles = st.last_n_candles.get(cfg.INDEX_SYMBOL, [])
        bn         = bn_candles[-1] if bn_candles else None

        market_open  = is_market_open()
        time_ok      = is_in_time_window()
        no_active    = st.active_trade is None
        elapsed      = time.monotonic() - st.last_exit_time if st.last_exit_time else float("inf")
        cooldown_ms  = elapsed * 1000

        side_range = sideways_range(bn_candles)
        range_ok   = side_range is not None and side_range >= 12
        mom        = strong_momentum(bn_candles, interval)
        leader_sig = leaders_momentum()
        sig_ok     = leader_sig.signal in ("BUY", "SELL")

        green = red = strong_qty = 0
        stock_stats = []
        for stock_name in cfg.LEADER_STOCKS:
            stock = next((s for s in cfg.STOCKS if s.name == stock_name), None)
            if not stock:
                continue
            sc        = st.last_n_candles.get(stock.symbol, [])
            c         = sc[-1] if sc else None
            if c:
                if   c.close > c.open: green += 1
                elif c.close < c.open: red   += 1
            qty       = st.latest_minute_qty.get(stock_name, 0.0)
            threshold = effective_threshold(stock_name, interval)
            if qty >= threshold:
                strong_qty += 1
            stock_stats.append(StockStat(stock=stock_name, candle=c, qty=qty, threshold=threshold))

        dir_ok          = max(green, red) >= cfg.SAME_DIRECTION_REQUIRED
        sq_ok           = strong_qty >= cfg.SAME_DIRECTION_REQUIRED
        candle_close_ok = bn is not None and _is_candle_closed(bn.start_time, interval)
        already_traded  = bn is not None and bn.start_time[:16] == st.last_trade_candle

        bn_ind = check_bn_indicators()
        st.bn_indicators = bn_ind

        macd_met = bn_ind.macd_dir not in (None, "—", "NEUTRAL")
        ema_met  = bn_ind.ema_stack is not None and (bn_ind.ema_stack.bullish or bn_ind.ema_stack.bearish)
        gate_ok  = bn_ind.bullish or bn_ind.bearish

        candle_close_time: Optional[str] = None
        if bn and bn.start_time:
            try:
                s     = bn.start_time[:19].replace(" ", "T")
                start = datetime.fromisoformat(s).replace(tzinfo=IST)
                mins  = {"3m": 3, "5m": 5, "15m": 15}.get(interval, 1)
                candle_close_time = (start + timedelta(minutes=mins)).strftime("%H:%M")
            except Exception:
                pass

        diag = EntryDiagnostics(
            market_open=market_open, time_window_ok=time_ok, no_active_trade=no_active,
            cooldown_ms=cooldown_ms, sideways_range=side_range, candle_close_ok=candle_close_ok,
            leader_signal_type=leader_sig.signal, leader_signal_reason=leader_sig.reason,
            green=green, red=red, strong_qty=strong_qty,
            already_traded_candle=already_traded, bn_ind=bn_ind, bn_candle=bn,
            stocks=stock_stats, momentum=MomResult(ok=mom.ok, reason=mom.reason),
            candle_close_time=candle_close_time,
        )
        st.entry_diagnostics = diag

        all_but_close = (
            market_open and time_ok and no_active
            and cooldown_ms >= 60_000 and range_ok and mom.ok
            and sig_ok and dir_ok and sq_ok and not already_traded
            and macd_met and ema_met and gate_ok
        )
        all_ok = all_but_close and candle_close_ok

        if all_but_close and not candle_close_ok:
            st.pending_signal = PendingSignal(type=leader_sig.signal, reason=leader_sig.reason)
        else:
            st.pending_signal = None

        if all_ok and st.active_trade is None:
            entry_price = st.bn_ltp if st.bn_ltp > 0 else (bn.close if bn else 0.0)
            candle_time = bn.start_time[:16] if bn else ""
            if bn_ind.bullish and leader_sig.signal == "BUY":
                await enter_trade(st, "BUY",  entry_price, candle_time, "AUTO", db)
            elif bn_ind.bearish and leader_sig.signal == "SELL":
                await enter_trade(st, "SELL", entry_price, candle_time, "AUTO", db)

        return diag


# ── Enter / Exit ──────────────────────────────────────────────────────────────

async def enter_trade(
    state: AppState,
    trade_type: str,
    price: float,
    candle_time: str,
    confidence: str,
    db: "DatabaseService | None" = None,
) -> None:
    if state.active_trade is not None:
        return
    num_lots             = _calc_num_lots(state)
    state.active_trade   = ActiveTrade(
        type=trade_type, entry=price, entry_time=candle_time,
        confidence=confidence, num_lots=num_lots,
    )
    state.last_trade_candle = candle_time
    state.signal_locked     = True

    t            = Trade()
    t.type       = trade_type
    t.price      = price
    t.time       = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    t.confidence = confidence
    t.pnl        = 0.0
    if db:
        await db.save_trade(t)
    print(f"ENTRY {trade_type} @ {price:.2f} [{candle_time}] conf={confidence} lots={num_lots}")


async def check_exit(state: AppState, db: "DatabaseService | None" = None) -> None:
    if state.active_trade is None:
        return
    bn_candles = state.last_n_candles.get(cfg.INDEX_SYMBOL, [])
    if not bn_candles:
        return
    price = state.bn_ltp if state.bn_ltp > 0 else bn_candles[-1].close
    at    = state.active_trade

    if at.type == "BUY":
        pnl = price - at.entry
        if   pnl >= cfg.TRAIL_TRIGGER:     at.current_sl = max(at.current_sl, price - cfg.TRAIL_DISTANCE)
        elif pnl >= cfg.BREAKEVEN_TRIGGER: at.current_sl = max(at.current_sl, at.entry)
        if price >= at.entry + cfg.TARGET or price <= at.current_sl:
            await _exit_trade(state, price, pnl, db)
    else:
        pnl = at.entry - price
        if   pnl >= cfg.TRAIL_TRIGGER:     at.current_sl = min(at.current_sl, price + cfg.TRAIL_DISTANCE)
        elif pnl >= cfg.BREAKEVEN_TRIGGER: at.current_sl = min(at.current_sl, at.entry)
        if price <= at.entry - cfg.TARGET or price >= at.current_sl:
            await _exit_trade(state, price, pnl, db)


async def _exit_trade(state: AppState, exit_price: float, pnl_pts: float,
                      db: "DatabaseService | None" = None) -> None:
    if state.active_trade is None:
        return
    pnl_pts  = round(pnl_pts, 2)
    num_lots = state.active_trade.num_lots
    pnl_rs   = round(pnl_pts * num_lots * cfg.LOT_SIZE, 2)

    t            = Trade()
    t.type       = state.active_trade.type + "_EXIT"
    t.price      = exit_price
    t.time       = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    t.confidence = state.active_trade.confidence
    t.pnl        = pnl_rs
    if db:
        await db.save_trade(t)

    state.available_funds += pnl_rs
    state.active_trade    = None
    state.last_exit_time  = time.monotonic()
    state.signal_locked   = False
    state.pending_signal  = None
    print(f"EXIT @ {exit_price:.2f} PnL={pnl_pts:.2f} pts = ₹{pnl_rs:.2f} ({num_lots} lots)")


# ── Manual entry / exit ───────────────────────────────────────────────────────

async def manual_entry(state: AppState, trade_type: str, price: float,
                       db: "DatabaseService | None" = None) -> None:
    if state.active_trade is not None:
        return
    candle_time = now_ist().strftime("%Y-%m-%dT%H:%M")
    await enter_trade(state, trade_type, price, candle_time, "MANUAL", db)


async def manual_exit(state: AppState, db: "DatabaseService | None" = None) -> None:
    if state.active_trade is None:
        return
    bn_candles = state.last_n_candles.get(cfg.INDEX_SYMBOL, [])
    if not bn_candles:
        return
    price = bn_candles[-1].close
    pnl   = (price - state.active_trade.entry
             if state.active_trade.type == "BUY"
             else state.active_trade.entry - price)
    await _exit_trade(state, price, pnl, db)
